internal/delegate/delegate.go

```diff
@@ -0,0 +1,139 @@
+//go:build windows
+
+package delegate
+
+import (
+	"sync"
+	"syscall"
+	"unsafe"
+
+	"github.com/go-ole/go-ole"
+)
+
+// Only a limited number of callbacks may be created in a single Go process,
+// and any memory allocated for these callbacks is never released.
+// Between NewCallback and NewCallbackCDecl, at least 1024 callbacks can always be created.
+var (
+	queryInterfaceCallback = syscall.NewCallback(queryInterface)
+	addRefCallback         = syscall.NewCallback(addRef)
+	releaseCallback        = syscall.NewCallback(release)
+	invokeCallback         = syscall.NewCallback(invoke)
+)
+
+// Delegate represents a WinRT delegate class.
+type Delegate interface {
+	GetIID() *ole.GUID
+	Invoke(instancePtr, rawArgs0, rawArgs1, rawArgs2, rawArgs3, rawArgs4, rawArgs5, rawArgs6, rawArgs7, rawArgs8 unsafe.Pointer) uintptr
+	AddRef() uint64
+	Release() uint64
+}
+
+// Callbacks contains the syscalls registered on Windows.
+type Callbacks struct {
+	QueryInterfaceCallback uintptr
+	AddRefCallback         uintptr
+	ReleaseCallback        uintptr
+	InvokeCallback         uintptr
+}
+
+var mutex = sync.RWMutex{}
+var instances = make(map[uintptr]Delegate)
+
+// RegisterCallbacks adds the given pointer and the Delegate it points to to our instances.
+// This is required to redirect received callbacks to the correct object instance.
+// The function returns the callbacks to use when creating a new delegate instance.
+func RegisterCallbacks(ptr unsafe.Pointer, inst Delegate) *Callbacks {
+	mutex.Lock()
+	defer mutex.Unlock()
+	instances[uintptr(ptr)] = inst
```

Comment:
Just asking but I'm a little bit worried about this as we can probably fall in weird Go GC issues? Like uintptr has a value but doesn't point to the value being claimed? 🤔

Reply:
This is called from the autogenerated code, and the pointer is always the result of calling `malloc`, so no GC involved. The only problem I can think of is that `malloc` could fail and return zero, but in that case the application would panic immediately.

Reply:
Also the release is captured here too, so we remove the pointer when an object is released.